Drop broken and superseded CiliumInternalIP restoration logic

giorio94 opened the pull request:
This old logic does work only when the IPAM is configured in clusterpool mode, as in all other cases we attempt the restoration from a plain k8s node resource, which does not contain the CiliumInternalIP information.

Hence, let's drop this obsolete logic, which is now superseded by the one introduced in c67d4eeea73c.

Moreover, the EnableHostIPRestore config option was introduced more than five years ago in a627e898067e65518a00306d795be34ce6272616 to allow disabling the CiliumInternalIP restoration logic in unit tests and prevent flakes caused by possible preexisting state.

Yet, this is no longer necessary, given that the relevant tests are already configured to use a temporary state directory, and in any case they are not expected to be run on a machine where another Cilium instance is running. Additionally, the k8s restoration logic recently introduced in c67d4eeea73c9d192776afab6ebc4f6107a21dc3 also didn't account for this flag.

To keep things simple, let's just drop this option and the remaining usages (affecting tests only).

Related: https://github.com/cilium/cilium/issues/29019, https://github.com/cilium/cilium/pull/29460

/cc @tommyp1ckles, @gandro

giorio94 committed:
node: drop broken CiliumInternalIP restoration

This old logic does work only when the IPAM is configured in clusterpool
mode, as in all other cases we attempt the restoration from a plain k8s
node resource, which does not contain the CiliumInternalIP information.

Hence, let's drop this obsolete logic, which is now superseded by the
one introduced in c67d4eeea73c ("daemon: use local CiliumNode resource
to populate CiliumInternalIP.")

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 commented on daemon/k8s/init.go:
```diff
@@ -142,29 +141,3 @@ func WaitForNodeInformation(ctx context.Context, log logrus.FieldLogger, localNo
-// Note that it does not validate the correctness of the IPs, as that is done
-// later in the daemon initialization when node.AutoComplete() is called.
-func restoreRouterHostIPs(n *nodeTypes.Node, log logrus.FieldLogger) {
-	if !option.Config.EnableHostIPRestore {
```
Here, the restoration was guarded by `EnableHostIPRestore` (and it consistent with the behavior of [ExtractCiliumHostIPFromFS](https://github.com/giorio94/cilium/blob/mio%2Fdrop-broken-cilium-internal-ip-restoration/pkg/node/address.go#L484-L487)). Instead, that parameter is no longer considered by the [new logic](https://github.com/giorio94/cilium/blob/mio%2Fdrop-broken-cilium-internal-ip-restoration/daemon/cmd/local_node_sync.go#L178-L189). Should we reintroduce it also there? Or just deprecate the parameter altogether?

gandro replied:
The rationale of the commit which introduced the flag six years ago says it mainly exists so unit tests don't trip up if they are being run from a developer environment: https://github.com/giorio94/cilium/commit/a627e898067e65518a00306d795be34ce6272616

That was when most Cilium developers were running cilium-agent directly on the host, not within a container. I think that mode of development has fallen out of favor. Therefore, I'd say let's deprecate and remove it (assuming it passes CI).

giorio94 replied:
I double-checked the usages, and the relevant codepaths are already configuring a temporary state directory to prevent interference in case of leftover state. There would still be the possibility of issues in case a `cilium_host` interface exists locally, but my feeling is that there would be many more tests breaking if another Cilium instance was running on the same host in any case.

Given that, I've just adopted the easiest solution and dropped that config option altogether (we don't even need the deprecation process, as it was not configurable by users in any ways).

@gandro PTAL

gandro replied:
I didn't even realize the flag wasn't exposed via CLI flags. That makes the case for removal even easier.

giorio94 replied:
Yep, neither did I.

gandro reviewed (approved):
Thanks!

giorio94 committed:
test: drop the EnableHostIPRestore config option

The EnableHostIPRestore config option was introduced more than five
years ago in a627e898067e ("agent: Correctly restore router IPs from
cilium_host interface") to allow disabling the CiliumInternalIP
restoration logic in unit tests and prevent flakes caused by possible
preexisting state.

Yet, this is no longer necessary, given that the relevant tests are
already configured to use a temporary state directory, and in any
case they are not expected to be run on a machine where another Cilium
instance is running. Additionally, the k8s restoration logic recently
introduced in c67d4eeea73c ("daemon: use local CiliumNode resource to
populate CiliumInternalIP.") also didn't account for this flag.

To keep things simple, let's just drop this option and the remaining
usages (affecting tests only).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>
